### ISSUE TITLE
Proposal: add `skip_local_storage` bool option to `remote_read` config

### ISSUE THREAD

**Issue by valyala** [prometheus/prometheus#4456]
### The issue
Sometimes remote read API endpoint contains all the data including the most recent data. Then there is no need to merge the data from local tsdb with the data read from remote read API endpoint, since the remote data fully covers all the local data. So reading local data for the subsequent merge and complete deduplication with the remote data becomes redundant and useless.

### The solution
It would be great to have `skip_local_storage` bool option in the <a href="https://prometheus.io/docs/prometheus/latest/configuration/configuration/#%3Cremote_read%3E">remote_read</a> config section, which could disable reading local data when the corresponding remote read API endpoint is in use.

**Comment by brian-brazil** [comment]
Can you explain more about your use case?

We have thus far explicitly not included an option to disable local storage. For reliability you always want to be able to use the data you have locally.

**Comment by valyala** [comment]
If multiple Prometheus instances (or other third-party scrapers) scrape distinct targets and write the collected data to a single remote storage, then each Prometheus instance is able to query all the data from the remote storage written by all the Prometheus instances, including the data missing in the local storage. In this case the merge of the remote data with the per-instance local data has little sense from the reliability point of view, since each Prometheus instance could return only a small part of the collected data, which is available in its' local storage, if the remote endpoint becomes unavailable.

Locally stored timeseries may clash unexpectedly with the timeseries written by another scrapers (such as Prometheus instances) into the remote storage. This may result in incorrect merge and, consequently, incorrect graphs.

The remote storage may transform its' data in various ways, which may break the subsequent merge of the data with Prometheus' locally stored data. For instance, the remote storage may reduce timeseries precision / resolution by aggregating the data over bigger intervals. This breaks `rate` graphs, since certain close points from remote storage and from local storage may have significantly varying values due to aggregation.

**Comment by brian-brazil** [comment]
> Locally stored timeseries may clash unexpectedly with the timeseries written by another scrapers (such as Prometheus instances) into the remote storage. This may result in incorrect merge and, consequently, incorrect graphs.

This is not a sane setup. Remote storage is meant to be an extension of local storage, not to replace it.

**Comment by brian-brazil** [comment]
To expand a bit, Prometheus is explicitly not meant to be a clustered system as that has complexity and reliability implications. Accordingly we aren't going to add features to allow users to try and make it one, as that'd end us up in the situation we want to avoid.

You should probably look at Cortex rather than trying to create one of these yourself.

**Comment by valyala** [comment]
> To expand a bit, Prometheus is explicitly not meant to be a clustered system as that has complexity and reliability implications. Accordingly we aren't going to add features to allow users to try and make it one, as that'd end us up in the situation we want to avoid.

The clustered system is already there - almost all the [existing remote storage integrations](https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage) are clustered in the reality. But it remains outside the Prometheus. The Prometheus just pushes data into these system and may query it back via remote storage API.

> You should probably look at Cortex rather than trying to create one of these yourself.

There is also [Thanos](https://github.com/improbable-eng/thanos), which provides global query view across all the Prometheus servers. I suppose it achieves this via remote storage API, so it may hurt from issues outlined above. cc'ing @Bplotka and @fabxc - `Thanos` developers for their clarifications on this question.

**Comment by brian-brazil** [comment]
> The Prometheus just pushes data into these systems and may query it back via remote storage API.

Yes, but always using the local storage as the primary.

> so it may hurt from issues outlined above

It works in a different way.

**Comment by Bplotka** [comment]
> There is also Thanos, which provides global query view across all the Prometheus servers. I suppose it achieves this via remote storage read API, so it may hurt from issues outlined above. 

Thanos does not use any of remote read/write features, except maybe direct usage of `/api/v1/read` protobuf enabled endpoint to fetch data required for global query. Actually local storage is extremely important for Thanos, since we use it as local persistent metric buffer. You can look in details how Thanos is designed here: https://improbable.io/games/blog/thanos-prometheus-at-scale (:

**Comment by henridf** [comment]
Fwiw I have one use case for this - using a Prometheus server solely as a query engine that reads from long-term storage via RR. This server doesn't do any scraping and so it doesn't write any samples to TSDB (unless I'm missing something). 

In this setup, it would be convenient to disable local storage altogether. But I do understand that this is far enough from the regular path that a fair answer might just be "build your own server using PromQL and relevant bits as libraries".

**Comment by danni-m** [comment]
We  would like to have the ability to use Prometheus as a reader without using the local storage (as @henridf suggested).

Also, remote read from a clustered (with HA) database can solve issues when you Prometheus server is down and you are bootstrapping a new one, it really makes sense to read the short-term data from remote.

**Comment by drewhemm** [comment]
Isn't this satisfied by the `read_recent: false` option for `remote_read` ?

**Comment by kamelkev** [comment]
@valyala how did you end up resolving your issue? Something within VictoriaMetrics? Or did you find some configuration operations that somehow ameliorated the problem?

**Comment by valyala** [comment]
@kamelkev , I just implemented PromQL engine from scratch inside VictoriaMetrics, so Grafana or other [Prometheus query API](https://prometheus.io/docs/prometheus/latest/querying/api/) clients may talk directly to VictoriaMetrics without Prometheus assistance.

**Comment by jiacai2050** [comment]
@drewhemm `read_recent: false` disable remote read, not local.

**Comment by drewhemm** [comment]
@jiacai2050 My bad, I got it the wrong way around. If remote storage is in use `read_recent: true` prevents Prometheus from reading data from the local storage to satisfy queries:

https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_read

It doesn't prevent it from _writing_ locally, but I believe it fulfils the requirement detailed by @valyala, i.e. having all Prometheus queries returned from remote storage.

Local storage retention could probably then be set really short, e.g. one day.

**Comment by jiacai2050** [comment]
> Local storage retention could probably then be set really short, e.g. one day.

It works, but kinds of ugly.
Since prometheus team have [no plan for long term storage](https://github.com/prometheus/docs/issues/149#issuecomment-127276802), I think it may provide `disable local tsdb` option to let users fully take advantage of remote storage.

**Comment by drewhemm** [comment]
@jiacai2050 I did quite a lot of experimenting with remote storage a few months back, using CrateDB as the backend; however, I observed that performing queries against remote storage is hugely inefficient compared with local storage, this is because (at least at the time) Prometheus retrieves a much larger data set from the remote than is actually returned as the result of the query. I believe this is by design, since the actual Prometheus query can only be executed locally. This in turn places a massive load on the remote storage for CPU, disk and network bandwidth. 

Because of this, I reverted to local storage and currently use AWS EFS instead to provide unlimited long-term multi-az storage, that also has the benefit of getting cheaper as the data gets older. I run Prometheus in Kubernetes as a Statefulset, which handles things like pod restarts etc.
